### ISSUE TITLE
Optimization ideas

### ISSUE THREAD

**Issue by newpavlov** [dgriffen/flif.rs#23]
- [ ] The biggest one is to use vector based storage and indices instead of pointers for maniac tree
- [ ] Currently boundary condition checks (`x!=0`, `y > 1`, etc.) are performed for every pixel, while most of the pixel do not need such checks. We could slightly modify pixel pass by dividing it into "boundary" and "core" passes
- [ ] Transformations use dynamic dispatch for no reason, enum for transformation variants which stores associated transformation data and vector of such enums should be more than enough.
- [ ] Maybe we could utilize explicit SIMD in some places?

**Comment by dgriffen** [comment]
As for explicit simd I'm not sure how we would approach that. All the operations tend to be on small sets of data and the flif standard unfortunately makes each channel of each pixel depend on previously decoded data.

**Comment by newpavlov** [comment]
Yes, I agree about maniac. Regarding ugly matches you can use the following approach to make code a bit nicer:
```Rust
// T1, T2, etc. transformations types defined in separate modules
// each type will implement `undo`, `range` and `crange` methods
enum Transforms {
    T1(T1),
    T2(T2),
    // ...
}

impl Transform {
    fn undo(&self, _pixel: &mut Pixel) {
        // can be automated with macro
        match self {
            Transforms::T1(t) => t.undo(),
            Transforms::T2(t) => t.undo(),
            // ...
        }
    }
}
```

Do you'll have just 3 match arms, on for each method.

As for SIMD, I've hoped there will be a prior art from FLIF team, but short search didn't produce any results...

**Comment by dgriffen** [comment]
Looks like there is some SIMD going on in the main repo, although I haven't dug in far enough to see what its used for primarily.

**Comment by newpavlov** [comment]
Added memory optimization to the list.

**Comment by newpavlov** [comment]
Another idea: we could drop generality over `Read` trait and instead assume that file is fully loaded into memory or memmap'ed. This will allow us to remove `Result` in several places (most importantly in `Rac` and consequently in pixels processing), which may improve performance due to the absence of branching. Ideally compiler should be able to deduce lack of errors if user passes `&[u8]` and eliminate branching, but I am not sure if it's able to do so. Probably it's worth to experiment with it.

**Comment by dgriffen** [comment]
I thought about that option. A better model may be to use some sort of push model for data so we can be given a partial buffer at any given time. That would allow us to progressively decode when we start supporting interlaced images
